shinclude.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SHINCLUDE" "1" "June 2018" "" ""
.
.SH "NAME"
\fBshinclude\fR \- include directives for code/markup comments
.
.SH "SYNOPSIS"
\fBshinclude\fR [options\.\.\.] \fIfile\fR
.
.br
\fBshinclude\fR [options\.\.\.] \fI\-\fR
.
.br
.
.SH "OPTIONS"
.
.SS "\-h, \-\-help"
help
.
.SS "\-V, \-\-version"
Show version
.
.SS "\-i, \-\-inplace"
Edit the file in\-place
.
.SS "\-p, \-\-shinclude\-path PATH"
Add path to path to look for \fBINCLUDE\fR and \fBRENDER\fR\.
.
.P
Can be repeated to add multiple paths\.
.
.P
Default: \fB("$PWD")\fR
.
.SS "\-c, \-\-comment\-style COMMENT_STYLE"
Comment style\. See \fICOMMENT STYLES\fR\.
.
.P
Use \fB\-c list\fR to get a list of available styles\.
.
.SS "\-cs, \-\-comment\-start COMMENT_START"
 Comment start\. Overrides language\-specific comment start\.
.
.P
See \fICOMMENT STYLES\fR\.
.
.SS "\-csa, \-\-comment\-start\-alternative COMMENT_START_ALTERNATIVE"
 Alternative comment start\. Overrides language\-specific comment start\.
.
.P
Useful for vim folds
.
.P
See \fICOMMENT STYLES\fR\.
.
.SS "\-ce, \-\-comment\-end COMMENT_END"
 Comment end\. Overrides language\-specific comment end\.
.
.P
See \fICOMMENT STYLES\fR\.
.
.SS "\-d, \-\-info"
Enable debug logging (\fI\fB$LOGLEVEL=1\fR\fR)
.
.SS "\-dd, \-\-debug"
Enable trace logging (\fB$LOGLEVEL=2\fR)\.
.
.SS "\-ddd, \-\-trace"
Enable trace logging (\fB$LOGLEVEL=2\fR) and print every statement as it is executed\.
.
.SH "BLOCK DIRECTIVES"
Included content is fenced by comments that contain \fBBEGIN\-<block\-type>\fR and \fBEND\-<block\-type>\fR respectively\.
.
.P
Fencing comments \fBmust\fR start at the start of the line\.
.
.IP "" 4
.
.nf

<block\-type> ::= "EVAL" | "INCLUDE" | "RENDER" | "BANNER" | "MARKDOWN\-TOC"
<nl> ::= "\en"
<spc> ::= " "
<begin\-line> ::= <com\-start> <spc> BEGIN\-<block\-type> <block\-arg>+ <spc>?  <com\-end>?
<end\-line> ::= <com\-start> <spc> END\-<block\-type> <spc>? <com\-end>?
<directive> ::= <begin\-line> <nl> <content\-line>* <end\-line> <nl>
.
.fi
.
.IP "" 0
.
.SS "EVAL"
Evaluates the arguments as a shell expression\. \fBBE CAREFUL\fR
.
.P
EVAL Runs on \fBfirst\fR pass
.
.IP "" 4
.
.nf

# BEGIN\-EVAL wc *
# END\-EVAL
.
.fi
.
.IP "" 0
.
.P
will be transformed to
.
.IP "" 4
.
.nf

# BEGIN\-EVAL wc *

  21   171  1085 LICENSE
  51   106   978 Makefile
 290   461  4058 README\.md
 558  1267 12212 shinclude
 275   828  5565 shinclude\.1
1723  4100 36080 total

# END\-EVAL
.
.fi
.
.IP "" 0
.
.P
Wrap in lines\. E\.g \fB\-w \'<pre>\' \'</pre\'\fR
.
.SS "INCLUDE"
Include data from a file\.
.
.P
\fBINCLUDE\fR runs on \fBfirst\fR pass\.
.
.IP "" 4
.
.nf

# BEGIN\-INCLUDE LICENSE
# END\-INCLUDE
.
.fi
.
.IP "" 0
.
.P
will be transformed to
.
.IP "" 4
.
.nf

# BEGIN\-INCLUDE LICENSE
The MIT License (MIT)

Copyright (c) 2016 Konstantin Baierer

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software\.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT\. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE\.
# END\-INCLUDE
.
.fi
.
.IP "" 0
.
.SS "RENDER"
Renders a file to markdown using a \fBshrender\fR \fIhttps://github\.com/kba/shrender\fR\.
.
.P
Runs on \fBfirst\fR pass
.
.SS "MARKDOWN\-TOC"
(source \fI\./src/block\-MARKDOWN\-TOC\.bash#L3\fR, test \fI\./test/MARKDOWN\-TOC\fR)
.
.P
Reads in the file and outputs a table of contents of the markdown headings\.
.
.P
Runs on \fBsecond\fR pass
.
.IP "" 4
.
.nf

# First Heading

[rem]: BEGIN\-MARKDOWN\-TOC
[rem]: END\-MARKDOWN\-TOC

## Second\-Level Heading
.
.fi
.
.IP "" 0
.
.P
will be transformed to (\fBshinclude \-cs \'[rem]:\' \-ce \'\' \-\fR)
.
.IP "" 4
.
.nf

# First Heading

[rem]: BEGIN\-MARKDOWN\-TOC

* [First Heading](#first\-heading)
    * [Second\-Level  Heading](#second\-level\-heading)

[rem]: END\-MARKDOWN\-TOC

## Second\-Level Heading
.
.fi
.
.IP "" 0
.
.P
String to indent a single level\. Default: \fB\fR
.
.P
Regex used to detect and tokenize headings\.
.
.P
Default: \fB^(##+)\es*(\.*)\fR
.
.P
Replace markdown links in heading with just the link text Link text: Remove \fB[ ]\fR Indentation: Concatenate \fB$MARKDOWN_TOC_INDENT\fR times the number of leading \fB#\fR\- 2
.
.SH "COMMENT STYLES"
.
.SS "xml"
Comment style:
.
.IP "" 4
.
.nf

  <!\-\- BEGIN\-\.\.\. \-\->
  <!\-\- END\-\.\.\. \-\->
.
.fi
.
.IP "" 0
.
.P
File Extensions:
.
.IP "\(bu" 4
\fB\.html\fR
.
.IP "\(bu" 4
\fB*\.xml\fR
.
.IP "" 0
.
.SS "markdown"
Comment style:
.
.IP "" 4
.
.nf

[]: BEGIN\-\.\.\.
[]: END\-\.\.\.
.
.fi
.
.IP "" 0
.
.P
Extensions: * \fB*\.ronn\fR * \fB*\.md\fR
.
.SS "pound"
Comment style:
.
.IP "" 4
.
.nf

# BEGIN\-\.\.\.
# END\-\.\.\.
.
.fi
.
.IP "" 0
.
.P
Extensions:
.
.IP "\(bu" 4
\fB*\.sh\fR
.
.IP "\(bu" 4
\fB*\.bash\fR
.
.IP "\(bu" 4
\fB*\.zsh\fR
.
.IP "\(bu" 4
\fB*\.py\fR
.
.IP "\(bu" 4
\fB*\.pl\fR
.
.IP "\(bu" 4
\fB*\.PL\fR
.
.IP "\(bu" 4
\fB*\.coffee\fR
.
.IP "" 0
.
.SS "slashstar"
Comment style:
.
.IP "" 4
.
.nf

/* BEGIN\-\.\.\. */
/* END\-\.\.\. */
.
.fi
.
.IP "" 0
.
.P
File Extensions:
.
.IP "\(bu" 4
\fB*\.cpp\fR
.
.IP "\(bu" 4
\fB*\.cxx\fR
.
.IP "\(bu" 4
\fB*\.java\fR
.
.IP "" 0
.
.SS "doubleslash"
File Extensions:
.
.IP "" 4
.
.nf

// BEGIN\-\.\.\.
// END\-\.\.\.
.
.fi
.
.IP "" 0
.
.P
File Extensions:
.
.IP "\(bu" 4
\fB*\.c\fR
.
.IP "\(bu" 4
\fB*\.js\fR
.
.IP "" 0
.
.SS "doublequote"
Comment style:
.
.IP "" 4
.
.nf

" BEGIN\-\.\.\.
" END\-\.\.\.
.
.fi
.
.IP "" 0
.
.P
File Extensions:
.
.IP "\(bu" 4
\fB*\.vim\fR
.
.IP "" 0
.
.SS "doubleslashbang"
Comment style:
.
.IP "" 4
.
.nf

//! BEGIN\-\.\.\.
//! END\-\.\.\.
.
.fi
.
.IP "" 0
.
.P
Extensions:
.
.IP "\(bu" 4
\fB*\.jade\fR
.
.IP "\(bu" 4
\fB*\.pug\fR
.
.IP "" 0
.
.SS "Vim fold"
Comment style:
.
.IP "" 4
.
.nf

#{{{ BEGIN\-\.\.\.
#}}} END\-\.\.\.
.
.fi
.
.IP "" 0
.
.SH "AUTHOR"
Konstantin Baierer \fIhttps://github\.com/kba\fR
.
.SH "SEE ALSO"
\fBshrender(1)\fR, \fBfiglet(1)\fR, \fBshlog(1)\fR
.
.SH "COPYRIGHT"
The MIT License (MIT)
.
.P
Copyright (c) 2016\-2018 Konstantin Baierer
.
.P
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
.
.P
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software\.
.
.P
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT\. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE\.


src/shinclude.bash
#!/bin/bash

typeset -A BLOCK_PASS
typeset -a SHINCLUDE_PATH
SHINCLUDE_PATH=()
SHINCLUDE_PATH+=("$PWD")

#-----------------------------------
# BEGIN-INCLUDE src/logging.bash
# END-INCLUDE
# BEGIN-INCLUDE src/style.bash
# END-INCLUDE
# BEGIN-INCLUDE src/block-EVAL.bash
# END-INCLUDE
# BEGIN-INCLUDE src/block-INCLUDE.bash
# END-INCLUDE
# BEGIN-INCLUDE src/block-RENDER.bash
# END-INCLUDE
# BEGIN-INCLUDE src/block-MARKDOWN-TOC.bash
# END-INCLUDE
# BEGIN-INCLUDE src/block-BANNER.bash
# END-INCLUDE
#-----------------------------------

# ### usage
#
# Show usage
#
usage() {
  echo "Usage: $0 [-h] [-c COMMENT_STYLE] [-i] <action> [input-file]"
  [[ ! -z "$1" ]] && echo -e "\n!! $* !!\n"
  echo '
  Input File:

    If input-file is "-", read from STDIN.
    If no input-file is given, assume "./README.md"'
}

# shellcheck disable=2094
_parse_lines() {
    local pass line begin end blocktype blockargs block infile
    pass=$1
    infile=$2
    IFS=$'\n'
    while read -r line;do
      if [[ "$line" = "$COMMENT_START BEGIN-"* ]];then
          begin="$line"

          blocktype="${begin:${#COMMENT_START}:${#line}}"
          blocktype="${blocktype##*BEGIN-}"
          blocktype="${blocktype%$COMMENT_END}"
          blocktype="${blocktype%% *}"

          blockargs="${begin:${#COMMENT_START}:${#line}}"
          blockargs="${blockargs#*BEGIN-$blocktype}"
          blockargs="${blockargs# }"
          blockargs="${blockargs% $COMMENT_END}"

          block=""
          while
              read -r line \
              && [[ "$line" != "$COMMENT_START END-$blocktype"* \
              && "$line" != *"END-$blocktype $COMMENT_END"   ]]
          do
              printf -v block "%s\n%s" "$block" "$line"
          done
          if ! declare -f -F "_block_$blocktype">/dev/null;then
              _error "No such block type '$blocktype'"
          fi
          if [[ ${BLOCK_PASS[$blocktype]} != $pass ]];then
              _debug 1 "PASS $pass: SKIP '$blocktype' '$blockargs'"
              printf "%s%s\n%s\n" "$begin" "$block" "$line"
              continue;
          fi
          _debug 1 "PASS $pass: RUN $blocktype '$blockargs'"
          _debug 2 "PASS $pass: RUN $blocktype '$blockargs' '$block'"
          newblock="$("_block_$blocktype" "$blockargs" "$block" "$infile")"
          ret=$?
          if [[ $ret != 0 ]];then
              _error "$ret: Error running '$blocktype' '$blockargs'"
          fi
          printf "%s\n%s\n\n%s\n" "$begin" "$newblock" "$line"
      else
          _debug 2 "PASS $pass: (ign) $line"
          printf "%s\n" "$line"
      fi
    done < "$infile"
}

## ## OPTIONS
_parse_args() {

    while [[ "$1" != '-' && "$1" = -* ]];do
        case "$1" in
            -h)
                ## ### -h, --help
                ##
                ## help
                ##
                usage
                exit ;;
            -i|--inplace)
                ## ### -i, --inplace
                ##
                ## Edit the file in-place
                ##
                IN_PLACE=true 
                ;;
            -p|--shinclude-path)
                ## ### -p, --shinclude-path PATH
                ##
                ## Add path to path to look for `INCLUDE` and `RENDER`.
                ##
                ## Can be repeated to add multiple paths.
                ##
                ## Default: `("$PWD")`
                ##
                SHINCLUDE_PATH+=("$2"); shift
                ;;
            -c|--coment-style)
                ## ### -c, --comment-style COMMENT_STYLE
                ##
                ## Comment style. See [COMMENT STYLES](#comment-styles).
                ##
                COMMENT_STYLE="$2"; shift
                ;;
            -cs|--comment-start)
                ## ### -cs, --comment-start COMMENT_START
                ##
                ## Comment start. Overrides language-specific comment start.
                ##
                ## See [COMMENT STYLES](#comment-styles).
                ##
                COMMENT_START="$2"; shift
                ;;
            -ce|--comment-end)
                ## ### -ce, --comment-end COMMENT_END
                ##
                ## Comment end. Overrides language-specific comment end.
                ##
                ## See [COMMENT STYLES](#comment-styles).
                ##
                COMMENT_END="$2"; shift
                ;;
            -d|--info)
                ## ### -d, --info
                ##
                ## Enable debug logging ([`$LOGLEVEL=1`](#loglevel))
                ##
                # shellcheck disable=2034
                LOGLEVEL=1
                ;;
            -dd|--debug) 
                ## ### -dd, --debug
                ##
                ## Enable trace logging (`$LOGLEVEL=2`).
                ##
                # shellcheck disable=2034
                LOGLEVEL=2
                ;;
            -ddd|--trace) 
                ## ### -ddd, --trace
                ##
                ## Enable trace logging (`$LOGLEVEL=2`) and print every statement as it is executed.
                ##
                # shellcheck disable=2034
                LOGLEVEL=2
                set -x
                ;;
        esac
        shift
    done

    if [[ "$1" && "$1" == "-" ]] ; then infile=/dev/stdin
    elif [[ "$1"              ]] ; then infile="$1"
    elif [[ -e "README.md"    ]] ; then infile="README.md"
    else _error "No file or stdin passed"; fi

    if [[ $IN_PLACE && "$infile" = "/dev/stdin" ]];then
        usage "Cannot edit STDIN in-place"
        exit 1
    fi

    COMMENT_STYLE=${COMMENT_STYLE:-$(_detect_comment_style "$infile")}
    if [[ -z "$COMMENT_STYLE" ]];then
        _error "Unable to detect comment style."
    fi
    COMMENT_START=${COMMENT_START:-${COMMENT_STYLE_START[$COMMENT_STYLE]}}
    COMMENT_END=${COMMENT_END:-${COMMENT_STYLE_END[$COMMENT_STYLE]}}
    _debug 1 "COMMENT_STYLE=$COMMENT_STYLE"
    _debug 1 "COMMENT_START=$COMMENT_START"
    _debug 1 "COMMENT_END=$COMMENT_END"
    _debug 1 "infile=$infile"
}

#
# main
#

_parse_args "$@"

# first pass
tempfile1=$(mktemp --tmpdir)
_debug 1 "tempfile1=$tempfile1"
trap 'rm $tempfile1' EXIT INT TERM
_parse_lines 1 "$infile" > "$tempfile1"

# second pass
tempfile2=$(mktemp --tmpdir)
_debug 1 "tempfile2=$tempfile2"
trap 'rm $tempfile2' EXIT INT TERM
_parse_lines 2 "$tempfile1" > "$tempfile2"

if [[ $IN_PLACE ]];then
    cp "$tempfile2" "$infile"
else
    cat "$tempfile2"
fi
